docs/architecture/adr-016-social-upgrades.md

```diff
@@ -0,0 +1,170 @@
+would halt all nodes in the network. This could be as simple as:
+
+```go
+func BeginBlocker(k keeper.Keeper, ctx sdk.Context, _ abci.RequestBeginBlock) {
+   if ctx.BlockHeader().Height >= BombHeight {
+       panic("social consensus must be reached to pass the bomb height")
+   }
+   return
+}
+```
+
+We could also impose other similar restrictions elsewhere instead, such as voting to reject the block if the deadline height is `BombHeight` is reached.
+
+```go
+func (app *App) ProcessProposal(req abci.RequestProcessProposal) abci.ResponseProcessProposal {
+    if req.Header.Height >= consts.BombHeight {
+        logInvalidPropBlock(app.Logger(), req.Header, "bomb height reached")
+        return abci.ResponseProcessProposal{
+            Result: abci.ResponseProcessProposal_REJECT,
+        }
+    }
+    ...
+}
+```
+
+Most importantly, we would also need to add this bomb height to light clients. This way validators cannot just ignore the bomb height and continue to produce blocks. There are multiple different ways to do that, but one universal way to make sure that all nodes, including light, halt at the bomb height would be to include it in the header verification logic:
+
+```go
+// ValidateBasic performs stateless validation on a Header returning an error
+// if any validation fails.
+//
+// NOTE: Timestamp validation is subtle and handled elsewhere.
+func (h Header) ValidateBasic() error {
+    if h.Height >= consts.BombHeight {
+        return errors.New("bomb height reached or exceeded")
+    }
+    ...
+    return nil
+}
+```
+
+#### Halting the Node using Social Consensus
+
+We hope to perform most upgrades using mechanism that doesn't involve shutting down and switching binaries, but depending on changes to the code, this might be difficult or not desirable (note that single binary syncing would still work fine). In that case, we would still require a mechanism to halt all nodes that are running the old binary in a way that respects social consensus. We can do that using the existing functionality in the application. Below are the configs in app.toml that would allow node operators to pick a height to shutdown their nodes at.
+
+```toml
+# HaltHeight contains a non-zero block height at which a node will gracefully
+# halt and shutdown that can be used to assist upgrades and testing.
+#
+# Note: Commitment of state will be attempted on the corresponding block.
+halt-height = 0
+
+# HaltTime contains a non-zero minimum block time (in Unix seconds) at which
+# a node will gracefully halt and shutdown that can be used to assist upgrades
+# and testing.
+#
+# Note: Commitment of state will be attempted on the corresponding block.
+halt-time = 0
+```
+
+## Consequences
+
+If we adopt Option 2, then we will be able to remove token voting from the statemachine sooner rather than later. This is riskier in that we will not have the battle tested mechanism, but it will force future upgrades that respect social consensus.
```

Comment:
```suggestion
If we adopt Option 2, then we will be able to remove token voting from the state machine sooner rather than later. This is riskier in that we will not have the battle tested mechanism, but it will force future upgrades that respect social consensus.
```